supervisor/docker/addon.py

```diff
@@ -523,18 +523,42 @@ def _run(self) -> None:
                 BusEvent.HARDWARE_NEW_DEVICE, self._hardware_events
             )
 
+    def _update(
+        self, version: AwesomeVersion, image: str | None = None, latest: bool = False
+    ) -> None:
+        """Update a docker image.
+
+        Need run inside executor.
+        """
+        image = image or self.image
+
+        _LOGGER.info(
+            "Updating image %s:%s to %s:%s", self.image, self.version, image, version
+        )
+
+        # Update docker image
+        self._install(
+            version, image=image, latest=latest, need_build=self.addon.latest_need_build
+        )
+
+        # Stop container & cleanup
+        with suppress(DockerError):
+            self._stop()
```

Comment:
Fine to bugfix it like this. But once we should I guess work with the StoreAddon also on update and swap it then over if it was successfully installed/builded

Reply:
So the challenging part here is that a `need_build` option only makes sense in the `install` method in `DockerAddon`. It doesn't make sense in `DockerInterface` or any of the other subclasses of `DockerInterface`. That means I can't really call `super()._update(...)` here because that will never pass `need_build` to `_install`.

The alternative I considered was updating the cached info for the addon prior to update. That way if `need_build` changed, we would know during `DockerAddon._install`. But that creates a new problem - if the update fails we have to try and rollback the update to the metadata to the prior version.

I also considered having `DockerAddon._install` look at `StoreAddon` to see if a build is needed instead of `Addon` but that won't work on reboot. `StoreAddon` contains the information for the latest version not the installed version. On reboot if we are missing the image we could do the wrong thing if the addon is out of date.

This seemed like the best option I could think of. But I'm definitely open to ideas if you have another. I didn't like copying and pasting that code up from `DockerInterface._install` either.

Reply:
I don't get why it needs self._stop() as that gets executed from Addon later

Reply:
@pvizeli When? `AddonManager.update` just calls backup, update, image cleanup and then start:
https://github.com/home-assistant/supervisor/blob/30932a83f8dcbb6261823b468aeb6e1d28b4a1cd/supervisor/addons/__init__.py#L249-L297

And update for addons just calls either install or build, both of which just pull/make the new image and don't touch the container:
https://github.com/home-assistant/supervisor/blob/30932a83f8dcbb6261823b468aeb6e1d28b4a1cd/supervisor/docker/interface.py#L204-L276
https://github.com/home-assistant/supervisor/blob/30932a83f8dcbb6261823b468aeb6e1d28b4a1cd/supervisor/docker/addon.py#L542-L576

Currently `DockerInterface._update` tells docker to stop and remove the container after the image is installed:
https://github.com/home-assistant/supervisor/blob/30932a83f8dcbb6261823b468aeb6e1d28b4a1cd/supervisor/docker/interface.py#L500-L501

If we don't move that into `DockerAddon._update` then I'm not sure what does it. If I'm missing something please link it but I'm not seeing it starting from the REST API.

Reply:
got it